src/sagemaker/serve/builder/jumpstart_builder.py

```diff
@@ -234,6 +254,217 @@ def _build_for_tgi_jumpstart(self):
 
         self.pysdk_model.env.update(env)
 
+    def _tune_for_js(
+        self, multiple_model_copies_enabled: bool = False, max_tuning_duration: int = 1800
+    ):
+        """Tune for Jumpstart Models in Local Mode.
+
+        Args:
+            multiple_model_copies_enabled (bool): Whether multiple model copies serving is enable by
+                this ``DLC``. Defaults to ``False``
+            max_tuning_duration (int): The maximum timeout to deploy this ``Model`` locally.
+                Default: ``1800``
+        returns:
+            Tuned Model.
+        """
+        if self.mode != Mode.LOCAL_CONTAINER:
+            logger.warning(
+                "Tuning is only a %s capability. Returning original model.", Mode.LOCAL_CONTAINER
+            )
+            return self.pysdk_model
+
+        num_shard_env_var_name = "SM_NUM_GPUS"
+        if "OPTION_TENSOR_PARALLEL_DEGREE" in self.pysdk_model.env.keys():
+            num_shard_env_var_name = "OPTION_TENSOR_PARALLEL_DEGREE"
+
+        initial_env_vars = copy.deepcopy(self.pysdk_model.env)
+        admissible_tensor_parallel_degrees = _get_admissible_tensor_parallel_degrees(
+            self.js_model_config
+        )
+        if isinstance(admissible_tensor_parallel_degrees, int):
+            admissible_tensor_parallel_degrees = [admissible_tensor_parallel_degrees]
+        available_gpus = max(admissible_tensor_parallel_degrees)
+
+        benchmark_results = {}
+        best_tuned_combination = None
+        timeout = datetime.now() + timedelta(seconds=max_tuning_duration)
+        for tensor_parallel_degree in admissible_tensor_parallel_degrees:
+            if datetime.now() > timeout:
+                logger.info("Max tuning duration reached. Tuning stopped.")
+                break
+
+            sagemaker_model_server_workers = 1
+            if multiple_model_copies_enabled:
+                sagemaker_model_server_workers = math.floor(available_gpus / tensor_parallel_degree)
+
+            self.pysdk_model.env.update(
+                {
+                    num_shard_env_var_name: str(tensor_parallel_degree),
+                    "SAGEMAKER_MODEL_SERVER_WORKERS": str(sagemaker_model_server_workers),
+                }
+            )
+
+            logger.info(
+                "Trying tensor parallel degree: %s, model server workers: %s...",
+                tensor_parallel_degree,
+                sagemaker_model_server_workers,
+            )
+
+            try:
```

Comment:
Why have we not moved this out to a separate function?

This function is overloaded with multiple responsibilities
1. Checking model types
2.  Override env variables
3.  Predict
4a.  Benchmark , record metrics
4b.  Catch a bunch of exceptions

I think separating it out as 1,2 & 3, 4(a, b) will make the code simpler to debug and read through